### ISSUE TITLE
[QUESTION] Should StreamController use this much RAM?

### ISSUE THREAD

**Issue by LordManhattan** [StreamController/StreamController#123]
Fairly high RAM usage by StreamController. Is this caused by a memory leak, or is it by design? I have plenty of RAM so it's not affecting me personally, but I'm just wondering if it's by design or not. 

![Screenshot from 2024-05-31 13-47-23](https://github.com/StreamController/StreamController/assets/7291676/b82edfa8-cebb-4976-98f2-a9c85c2225af)

**Comment by Core447** [comment]
Yeah, high memory usage is still a big problem. I think it is related to the actions as the consumption grows when adding actions and switching pages. I hope I can fix this soon

**Comment by pnmtech** [comment]
I just realized to a much lesser extent that I have the same issue and with some short research I found m this https://github.com/flatpak/xdg-dbus-proxy/issues/51 and apparently the fix was found in https://github.com/sailfishos/xdg-dbus-proxy/blob/master/rpm/0001-Fix-GVariant-reference-leaks.patch I am not good with programming so I have no idea if my input was needed in the first place. I'm just trying to help. Really love the software. I've been enjoying using it

**Comment by Core447** [comment]
@pnmtech Thank you for the links! However, I think the issue is caused by the app itself, because the same problem appears when using it outside of Flatpak

**Comment by NightSprinter** [comment]
It isn't just the RAM usage, CPU as well, both Flatpak or otherwise.  Both get worse if you just make a button show the percentage of CPU or RAM used, and just press that button rapidly.  I also wonder if it's something in the underlying Python layer that StreamController uses?

**Comment by Core447** [comment]
> CPU or RAM

The graphs or the "normal" ones? Because the graphs take up a huge amount of memory (that's why I added a warning in their config area).

I hope I can fix those issues soon, but it'll probably take some time to find the issues

**Comment by Linuturk** [comment]
Wanted to chime in that I'm seeing this behavior as well, leading to resource contention on my system.

When you first start StreamController, memory usage is minimum.

![Image](https://github.com/user-attachments/assets/a130d49e-f95a-44ba-bf39-8cdc9516c118)

![Image](https://github.com/user-attachments/assets/7ab9f9ab-4b4b-4334-a7be-444d55369cad)

But over a day or two, you can see memory usage has exploded, and just gets worse over time.

![Image](https://github.com/user-attachments/assets/d797570a-96f4-4886-8fdd-a9d5b05f3aab)

**Comment by Phrozenn1** [comment]
1.8GB is my current record. I hope this get fixed soon

**Comment by Dragonsong3k** [comment]
> 1.8GB is my current record. I hope this get fixed soon


Out of curiousity, does your StreamDeck freeze up when the memory is high?

**Comment by ImDevinC** [comment]
FWIW: The screenshot from @Linuturk is not an issue with StreamController, but rather a bug with xdg-proxy as seen here: https://github.com/flatpak/xdg-dbus-proxy/issues/51. They have supposedly fix this, but have not released the fix yet

Some things that can help here:
1. What does the total memory landscape look like. Using atool like htop can make this easier to see, as it is helpful to see VIRT vs RES vs SHR memory (this is also shown in the original screenshot report from whatever app that is)
2. How is your StreamController configured?
  - How many pages do you have configured?
  - How many actions on the page?
  - What plugins are installed?

**Comment by Dragonsong3k** [comment]
Here is my load out if it helps. 

I have 1 page with various buttons from those plugins. 

The buttons I have configured
- MicMute
- Qemu VM Status / Toggle
- Basic Next / Previous / PlayPause
- Spofity Now Playing 
- Audio Output Switcher ( 2 buttons, each assigned to an output )
- Clock showing time

![Screenshot_20251015_130000_GitHub.jpg](https://github.com/user-attachments/assets/c09a21f6-c0e0-46e0-9f28-a2e8ca48415a)

From what I see, it's the xdg process and the Python process pulling about 3.5 GBs of ram after a full days usage. 

The StreamDeck+ will also lock up after about a day of usage. I am not 100% convinced it is related to the memory usage. 

I can get more details when I get back to my system.

**Comment by ImDevinC** [comment]
Thanks. The xdg-dbus-proxy is the issue I linked above, it is unfortunately not somethin we can fix and requires an upstream fix. That screenshot also doesn't really provide information if that memory column is virtual memory or resident memory. Not a huge deal, but would help in understanding how big of an issue this is (virtual memory is just how much memory has been requested by the application, now how much is actively being used)

**Comment by ImDevinC** [comment]
<img width="706" height="448" alt="Image" src="https://github.com/user-attachments/assets/4bcd2da7-49d6-49fa-a8f9-139a46c5fd91" />
I'm having a lot of trouble reproducing any issues that others are reporting so far. I've been running for about 5 hours and haven't seen a massive spike. I did see a small spike in usage when the system locked and a screensaver was put on the device, but I'm not seeing any huge massive memory usage spikes likes others are reporting. As you can see from the left side, this is <700MB, which I don't _think_ is too bad as we're using a lot of UI elements and having to communicate with a physical device.
But this is also why I'm asking for a breakdown of memory because `htop` actually shows me at almost 5G of VIRT, but only 659M of RES, and RES is the actual number that matters. This lines up with the first screenshot as well, showing that reserved memory is only ~800MB.

**Comment by LordManhattan** [comment]
A small update from me. It's currently using 438 MB of RAM (background usage with no GUI) with a system uptime of 4 days and 12 hours. Running the GUI, RAM usage jumps to 480 MB, which is _a lot_ lower than I expected. 

It does look like RAM usage is growing slowly but steadily when the GUI is open, though, going from 480 to 503 MB in about 20 minutes.  Closing the app (GUI) does _not_ free up that RAM  🤔

**EDIT:** Launching the app (GUI) _again_ causes the RAM usage to grow a little, but the app does not free up that RAM when closing the app (GUI). 

Plugins I use: 

- Mic Mute toggle
- Audio Output toggle
- Digital clock
- Media controls
- Brightness controls
- System suspend (command)

OS: Fedora 42 Workstation (Gnome).

**Comment by ImDevinC** [comment]
Thanks @LordManhattan, can you also describe what plugins and actions you have installed? Most of these are pretty self-explanatory, but Mic Mute and Audio Output are available in a couple different plugins

**Comment by ImDevinC** [comment]
If anyone can export their  pages they use and upload it here, that would also be helpful. There should be no sensitive information unless you are using a command that you put something sensitive in.

To do this:
1. In the upper left corner of the app, next to the Page dropdown, click the folder icon
2. Select any page
3. Click the three dots in the upper right corner and choose Export All
4. Save the JSON file somewhere
5. Upload the file here (or you can message me on Discord, I'm in the StreamController discord with the same username)

**Comment by LordManhattan** [comment]
Sure can do, @ImDevinC. Not a lot to look at here, though. 

[StreamController_LordManhattan.json](https://github.com/user-attachments/files/22956857/StreamController_LordManhattan.json)

**Comment by Phrozenn1** [comment]
> > 1.8GB is my current record. I hope this get fixed soon
> 
> Out of curiousity, does your StreamDeck freeze up when the memory is high?

It does, I made a script that restarts SC once a day to mitigate some of this. But I still don't think this is an application that should use ~1GB of ram with 1 page and 7 buttons. Even chromium uses less ram with 20 tabs open.

Attached the json file requested by @ImDevinC and screenshot of btop.

<img width="2041" height="1400" alt="Image" src="https://github.com/user-attachments/assets/f7e6bfa9-5635-4b40-a95b-54debe8984a1" />

[StreamController_2025-10-17_08-55.json](https://github.com/user-attachments/files/22965236/StreamController_2025-10-17_08-55.json)

**Comment by ImDevinC** [comment]
Thanks for the JSON files, this allows me to recreate your deck fairly close to what you have. Out of curiosity @Phrozenn1, how long was StreamController running in that scenario? And do you normally leave the GUI open, do you cnormally leave it closed, are you opening/closing it throughout the day? Trying to get usage information so I can try to replicate because I've never gotten anywhere close to your memory number and want to try and replicate.

**Comment by Phrozenn1** [comment]
@ImDevinC, I'd estimate between 24-36 hours. I did an update of Omarchy yesterday morning and it's 7.30pm now. I never open the GUI unless I need to tweak something or add another button.
Hardware wise it's a 5950X, 32GB G.skill ram, Gigabyte 4080S with 16GB ram.
I should probably just learn to code so I can actually contribute to projects. Instead of just complain when something's not right ;)

**Comment by ImDevinC** [comment]
https://limewire.com/d/T4Dts#2UbSLBsMpK

This isn't an official build, but I have put together a new Flatpak build of StreamController at the link above (it will expire after 1 week). You don't have to uninstall your existing StreamController to use it, but if you want to make sure the rihgt one launches you'll need to run StreamController from a terminal and use `flatpak run --branch=master com.core447.StreamController`. There are no breaking changes to the existing flow though, so uninstalling the old one will also work, and you can reinstall it later if you want. 

The full change list can be found here: https://github.com/StreamController/StreamController/pull/484. This does a lot to try and help manage memory. I did some local testing on my end, but getting other feedback will help a lot

*edit*
Forgot to mention, you can install this build by downloading it and then running `flatpak install StreamController.flatpak`

**Comment by Phrozenn1** [comment]
There was an update a few days that seems to have made the situation better. SC is hovering at about the 650MB mark constantly when the gui is not up, and it doesn't seem to add to the usage in a very significant way when launched. 
I still think 650MB is a lot for what I have configured now, but it's tolerable. For all I know this might be the nature of flatpaks or the way the hardware needs this type of application to work, but for comparison, qutebrowser with 15 tabs open is hovering around 700MB. 

I've also noticed the cpu usage is between 6-10% most of the time, even when the process is sleeping.

**Comment by ImDevinC** [comment]
@Phrozenn1 is this with the package I provided above? There hasn't been an update to StreamController in the past few days (1.5.0-beta.12 came out a couple weeks ago, but had no changes that should impact memory usage)

**Comment by phantomBit** [comment]
1.5.0-beta.12 most definitely still has this issue. just racked up 14GB of ram for xdg-dbus-proxy and flatpak-session-helper is using 6GB of ram.

running for days it just keeps eating until restart. attaching my config for context. ill disable the album info and weather since those constant updates seem to be some of the problem from what i have read here. 

<img width="699" height="495" alt="Image" src="https://github.com/user-attachments/assets/8719b9ca-e294-4031-9cef-944e62d9ef9a" />

```
{
    "brightness": {
        "value": 75
    },
    "background": {
        "overwrite": true,
        "path": null,
        "loop": true
    },
    "keys": {
        "0x0": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_Weather::Weather",
                            "settings": {
                                "unit": 2,
                                "lat": "00.00",
                                "lon": "00.00"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "background": {
                        "color": [
                            0,
                            0,
                            0,
                            0
                        ]
                    }
                }
            }
        },
        "1x0": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_OSPlugin::Hotkey",
                            "settings": {
                                "keys": [
                                    [
                                        29,
                                        1
                                    ],
                                    [
                                        56,
                                        1
                                    ],
                                    [
                                        105,
                                        1
                                    ],
                                    [
                                        105,
                                        0
                                    ],
                                    [
                                        56,
                                        0
                                    ],
                                    [
                                        29,
                                        0
                                    ]
                                ]
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": "/home/phantomBit/.var/app/com.core447.StreamController/data/icons/com_axolotlmaid_FontAwesomeIcons/icons/white/arrow-left.svg"
                    },
                    "background": {
                        "color": [
                            97,
                            53,
                            131,
                            255
                        ]
                    }
                }
            }
        },
        "4x0": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_OSPlugin::RunCommand",
                            "settings": {
                                "command": "youtube-music %U"
                            }
                        }
                    ],
                    "image-control-action": null,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": "/home/phantomBit/.var/app/com.core447.StreamController/data/Assets/AssetManager/Assets/ytmusic.png"
                    }
                }
            }
        },
        "4x2": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_MediaPlugin::PlayPause",
                            "settings": {
                                "show_label": false,
                                "show_thumbnail": false,
                                "player_name": "com.github.th_ch.youtube_music"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": null
                    }
                }
            }
        },
        "2x2": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_MediaPlugin::Previous",
                            "settings": {
                                "show_label": false,
                                "show_thumbnail": false,
                                "player_name": "com.github.th_ch.youtube_music"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0
                }
            }
        },
        "2x1": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_OSPlugin::Hotkey",
                            "settings": {
                                "keys": [
                                    [
                                        29,
                                        1
                                    ],
                                    [
                                        42,
                                        1
                                    ],
                                    [
                                        108,
                                        1
                                    ],
                                    [
                                        108,
                                        0
                                    ],
                                    [
                                        29,
                                        0
                                    ],
                                    [
                                        42,
                                        0
                                    ]
                                ]
                            }
                        }
                    ],
                    "image-control-action": null,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": "/home/phantomBit/.var/app/com.core447.StreamController/data/icons/com_core447_MaterialIcons/icons/volume_down-inv.png"
                    }
                }
            }
        },
        "1x1": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_gapls_AudioControl::AdjustVolume",
                            "settings": {
                                "volume-adjust": 5.0,
                                "pulse-name": "alsa_output.usb-Generic_USB_Audio-00.HiFi__hw_Audio__sink",
                                "device-filter": "sink",
                                "use-standard": true,
                                "show-info": true,
                                "show-device-name": false
                            },
                            "event-assignments": {
                                "Key Down": "adjust-volume-positive"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0
                }
            }
        },
        "0x1": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_gapls_AudioControl::AdjustVolume",
                            "settings": {
                                "volume-adjust": -5.0,
                                "info-content": "volume",
                                "pulse-name": "alsa_output.usb-Generic_USB_Audio-00.HiFi__hw_Audio__sink",
                                "use-standard": true,
                                "device-filter": "sink"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": null
                    }
                }
            }
        },
        "3x0": {
            "states": {
                "0": {
                    "actions": [],
                    "image-control-action": null,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": "/home/phantomBit/.var/app/com.core447.StreamController/data/icons/com_axolotlmaid_FontAwesomeIcons/icons/white/arrow-right.svg"
                    },
                    "background": {
                        "color": [
                            97,
                            53,
                            131,
                            255
                        ]
                    }
                }
            }
        },
        "3x2": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_MediaPlugin::Next",
                            "settings": {
                                "show_label": false,
                                "show_thumbnail": false,
                                "player_name": "com.github.th_ch.youtube_music"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0
                }
            }
        },
        "4x1": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_MediaPlugin::Info",
                            "settings": {
                                "show_label": true,
                                "show_thumbnail": true,
                                "seperator_text": "--"
                            }
                        }
                    ],
                    "image-control-action": 0,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0
                }
            }
        },
        "3x1": {
            "states": {
                "0": {
                    "actions": [
                        {
                            "id": "com_core447_OSPlugin::Hotkey",
                            "settings": {
                                "keys": [
                                    [
                                        29,
                                        1
                                    ],
                                    [
                                        42,
                                        1
                                    ],
                                    [
                                        103,
                                        1
                                    ],
                                    [
                                        103,
                                        0
                                    ],
                                    [
                                        42,
                                        0
                                    ],
                                    [
                                        29,
                                        0
                                    ]
                                ]
                            }
                        }
                    ],
                    "image-control-action": null,
                    "label-control-actions": [
                        0,
                        0,
                        0
                    ],
                    "background-control-action": 0,
                    "media": {
                        "path": "/home/phantomBit/.var/app/com.core447.StreamController/data/icons/com_core447_MaterialIcons/icons/volume_up-inv.png"
                    }
                }
            }
        }
    }
}
```

**Comment by ImDevinC** [comment]
1.5.0-beta.12 did not contain any fixes for high memory usage, so not surprised. At this point, we need people to test the build linked above (https://limewire.com/d/T4Dts#2UbSLBsMpK). Understandable if you don't trust it, but you can rebuild the flatpak on your own if you want and point it to my repo/branch (https://github.com/StreamController/StreamController/compare/main...ImDevinC:StreamController:memory-changes).

I also think that a large part of the high memory usage has to do with the background image. The build I linked above does include changes around how we manage background images, but would be nice to get confirmation from others if this helps at all

**Comment by phantomBit** [comment]
Sorry i did not mean i expected the latest to have fixed it, just noting the version i was on. Limewire is still a thing? crazy.

ill check out the build and see if it makes a difference. i pretty reliably get over 5GB ram usage with my setup so it would be nice to have this fixed.

**Comment by PLAY3R2130** [comment]
Hey all, 

just wanted to post i am seeing the same, had me looking at other Linux OS's trying to fix my Kernel Panic issues lol

XDG-dbus-proxy was at 4.5 GB RAM before i quit the StreamController then it dropped to 400 KB lol

Im on Ubuntu.

PRETTY_NAME="Ubuntu 24.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04.3 LTS (Noble Numbat)"
VERSION_CODENAME=noble

Edit, seems i have the Beta installed (1.5.0-beta.12) so, it hasn't helped, anything i can help with to get this fixed?

**Comment by ImDevinC** [comment]
@PLAY3R2130 You can install both, but it's easiest to uninstall the old and install the new one above. It won't mess up any settings and your can always uninstall the new one and install the official one again

**Comment by PLAY3R2130** [comment]
> [@PLAY3R2130](https://github.com/PLAY3R2130) You can install both, but it's easiest to uninstall the old and install the new one above. It won't mess up any settings and your can always uninstall the new one and install the official one again

I do in fact have the beta installed, whats next?

**Comment by ImDevinC** [comment]
Just to confirm, the regular version is beta, this other version is essentially an alpha but the version is the same. Once you have it installed, just need to use it like normal and check resource usage

**Comment by PLAY3R2130** [comment]
> Just to confirm, the regular version is beta, this other version is essentially an alpha but the version is the same. Once you have it installed, just need to use it like normal and check resource usage

I uninstalled the one i had installed via Software Manager and installed the one i downloaded via the above link. So far the memory usage is good, will need to test for a day or two to see how it does

**Comment by PLAY3R2130** [comment]
In the 6 hours I have been "testing", RAM usage for XDG-DBUS-Proxy started at 300KB and is now at 1 GB. After restarting the program, usage dropped back down to 300KB

Looks like the leak is still there, but it took longer to get up to a level i didn't like

**Comment by wafflesoup** [comment]
I've been running this build for a couple days now and flatpak-session-helper is still ballooning to using over 4GB of ram after a couple of days

**Comment by ImDevinC** [comment]
The issue with flatpak-session-helper and xdg-dbus-proxy is due to an issue in Flatpak, not StreamController directly. I've linked to that bug above. 
As for Streamcontroller memory usage, if you're still seeing it very high, can you try removing any wallpapers you have on the deck and see if that helps at all? I've noticed that wallpapers seem to have a large impact and would like to confirm

**Comment by PLAY3R2130** [comment]
> The issue with flatpak-session-helper and xdg-dbus-proxy is due to an issue in Flatpak, not StreamController directly. I've linked to that bug above. As for Streamcontroller memory usage, if you're still seeing it very high, can you try removing any wallpapers you have on the deck and see if that helps at all? I've noticed that wallpapers seem to have a large impact and would like to confirm

Just wanted to double check on this, but I only am using colors, no wallpapers

**Comment by PLAY3R2130** [comment]
Not sure why, but it seems the Memory Leak might be gone for me after my system updated to Ubuntu 25.04. 

xdg-dbus-proxy is sitting at 400 kb and has been for almost 12 hours now.

PRETTY_NAME="Ubuntu 25.04"
NAME="Ubuntu"
VERSION_ID="25.04"
VERSION="25.04 (Plucky Puffin)"
VERSION_CODENAME=plucky
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=plucky
LOGO=ubuntu-logo

**Comment by mike3nl** [comment]
@ImDevinC 
Hi Devin.

Little update or better info from my side.

Stream Deck + gets slower and yes i know the memory leak. 
- I'm on Linux Mint 22.2 Zara, for latest updates checked today
- Current_desktop XFCE
- Stream Controller 1.5.0-beta.12
- On the screenshot you can see that "python 3.12" is running up the memory use.
- Now it uses 734mb Has started as something by 200mb i mean and it can run up to 945mb.

Points this all to python? 
And there was an update on an external library which is not in this release, but can't remember right now. Will find it...
- It si the python-elgato-streamdeck where Core447 helped out too.
[https://github.com/abcminiuser/python-elgato-streamdeck](url)

- I have 4 sites with a menu start site. I do this because i'm from the automation industry and we work with menu's, haha.

All for information cases

Thanks and hold up the good work.
I'm not a python guy, i'm co

<img width="1276" height="395" alt="Image" src="https://github.com/user-attachments/assets/6bce84d0-03e7-49f2-ac1a-f38304a55900" />

ming from Ansi-C and C++.

**Comment by bitclick** [comment]
i am affected too, running current flatpak in xmonad (no gnome etc.).  My findings:
- Streamcontroller focus detection falls back to polling via xprop, which is expensive
  - runs xprop every 200ms ❗  https://github.com/StreamController/StreamController/blob/0ae82dde81afbdac5f52382fd15a1904d223491d/src/backend/WindowGrabber/Integrations/X11.py#L189
  - this is even more expensive inside the flatpak sandbox ❓ 
- i suspect, the cpu usage in `flatpak-session-helper` comes from this


mitigation via AI-consult:
**UPDATE: doesn't work after all - polling is still strong with this one**
> - i made sure, xmonad provieds ewmh hints
> - ran `flatpak override --user --socket=x11 --env=GDK_BACKEND=x11 com.core447.StreamController` to give StreamController direct access to read those hints  (pasted from gemini suggestions)
> - after killing `flatpak-session-helper` and restarting StreamController, it runs without `xprop`-polling and i have low cpu activity in `flatpak-session-helper`

Could we at the least get a user warning when the `xprop` polling gets used? and maybe make the interval configurable.  Maybe the permissions for the flatpak-release need to be updated.

UPDATE:
- looking through the code, it does not try to use of EWMH to query the active window: https://specifications.freedesktop.org/wm/latest/ar01s03.html#id-1.4.10
- multiple python libs are available, e.g. https://github.com/Kalmat/EWMHlib/blob/master/docstrings.md#getactivewindow

**Comment by ImDevinC** [comment]
Ok, I have a new way to test. Please remove this beta version and install the latest official release (1.5.0-beta.13). Flatpak would be preferred, but any method should be fine.

Once installed, go to Settings -> Store enable Custom Plugins and add a custom plugin:
URL: https://github.com/ImDevinC/StreamControllerDebugPlugin
Branch: main

Then close the settings and go the store. Find the Debug plugin (it's big and red) and install it. Then you should have a new action in the debug plugin called Profiler. Add that to a button, and then press the button one time to start profiling. Occasionally, when you notice RAM increases, press the button again to take a snapshot and it will be print some data to the log. If you want to stop profiling at any point, just long press the button. 

After some amount of time, whenever you think the jump is relevant, please grab the logs from ~/.var/app/com.core447.StreamController/data/logs/logs.log and upload them. You shold be able to keep StreamController running and grab logs occasionally to send them.

**Comment by Elip100** [comment]
Was having this same problem and came across this GitHub issue. I ran the profiler and took a snapshot once my memory usage hit 50% (over 8gb from StreamController). Here are the logs: https://www.mediafire.com/file/9blo9t1yuszd6lz/logs-original.log/file. There are hundreds of thousands of lines of the Spotify controller plugin complaining about something, so I also had a python script create a version with those lines removed: [logs-cleaned.log](https://github.com/user-attachments/files/25058018/logs-cleaned.log). Hope this helps in some way! Note: it took about a week to hit this memory usage.

**Comment by ImDevinC** [comment]
Welp, unfortunately that doesn't really show anything that indicates what is taking up that amount of memory. Only seeing a few hundred megabytes. 

```
2026-02-03 16:40:07.406 | DEBUG    | plugins.com_imdevinc_StreamControllerDebugPlugin.actions.DebugAction:_take_snapshot:58 - First snapshot memory usage differences:
2026-02-03 16:40:07.412 | DEBUG    | plugins.com_imdevinc_StreamControllerDebugPlugin.actions.DebugAction:_take_snapshot:59 - /app/lib/python3.12/site-packages/loguru/_logger.py:2016: size=510 MiB (+510 MiB), count=2302907 (+2302907), average=232 B
/app/lib/python3.12/site-packages/loguru/_handler.py:197: size=293 MiB (+293 MiB), count=2302908 (+2302908), average=134 B
/app/lib/python3.12/site-packages/loguru/_datetime.py:160: size=87.8 MiB (+87.8 MiB), count=1151454 (+1151454), average=80 B
/app/lib/python3.12/site-packages/loguru/_datetime.py:158: size=79.1 MiB (+79.1 MiB), count=2302908 (+2302908), average=36 B
/app/lib/python3.12/site-packages/loguru/_logger.py:2019: size=70.3 MiB (+70.3 MiB), count=1151453 (+1151453), average=64 B
<frozen posixpath>:174: size=67.0 MiB (+67.0 MiB), count=1151454 (+1151454), average=61 B
<frozen genericpath>:151: size=63.7 MiB (+63.7 MiB), count=1151454 (+1151454), average=58 B
/app/lib/python3.12/site-packages/loguru/_logger.py:2022: size=61.5 MiB (+61.5 MiB), count=1151454 (+1151454), average=56 B
/app/lib/python3.12/site-packages/loguru/_logger.py:2028: size=52.7 MiB (+52.7 MiB), count=1151454 (+1151454), average=48 B
/app/lib/python3.12/site-packages/loguru/_logger.py:2027: size=52.7 MiB (+52.7 MiB), count=1151454 (+1151454), average=48 B
```

There is an open PR for a resource leak that may have some impact, just waiting for it to be merged and released to see if it has any improvements: https://github.com/StreamController/StreamController/pull/527

**Comment by notmugi** [comment]
i made a basic controller for mpc /mpd with two backgrounds and two idle backgrounds and it ended up taking a total of 1.8gb of ram which i find pretty unacceptable. will test tomorrow after a reboot and see if it takes up less now that im done tinkering with it.

**Comment by notmugi** [comment]
> i made a basic controller for mpc /mpd with two backgrounds and two idle backgrounds and it ended up taking a total of 1.8gb of ram which i find pretty unacceptable. will test tomorrow after a reboot and see if it takes up less now that im done tinkering with it.

i felt obligated to check on this now instead of later, but i can report that the usage is lower when starting the app fresh. however, swapping pages always increases the memory usage despite never clearing said usage, so it’s always increasing. still not good. if i switch between the same two pages multiple times why would my memory usage need to be going from 300mb to 600? still really nasty memory management issues regardless.

**Comment by Elip100** [comment]
> i can report that the usage is lower when starting the app fresh.

Yes. The memory usage is always low when first opening the app or restarting it. However, over time the usage increases somewhat rapidly without stopping. It can gain multiple gigabytes in a week. Both `StreamController` and `flatpak-session-helper` have this issue.

**Comment by sifmelcara** [comment]
I'm running `1.5.0-beta.13` and StreamController uses 22GB of memory after 6 days

Have anyone tested https://github.com/StreamController/StreamController/pull/527 and can confirm if it fixes the leak or not?

If it's still not fixed I might be able to help look into it (figured I should ask here first before investing time running an unreleased version to repro and debug it..)

**Comment by Elip100** [comment]
> Have anyone tested [#527](https://github.com/StreamController/StreamController/pull/527) and can confirm if it fixes the leak or not?
> 
> If it's still not fixed I might be able to help look into it (figured I should ask here first before investing time running an unreleased version to repro and debug it..)

I've been testing it for a few days and still had memory issues

**Comment by notmugi** [comment]
if anyone has looked into [#558](https://github.com/StreamController/StreamController/pull/558) it looks very promising. haven’t tested the ram improvements myself but it seems to do quite a bit.

**Comment by sifmelcara** [comment]
> I'm running 1.5.0-beta.13 and StreamController uses 22GB of memory after 6 days
> If it's still not fixed I might be able to help look into it (figured I should ask here first before investing time running an unreleased version to repro and debug it..)

Just a quick update, I'm profiling `1.5.0-beta.14` right now but noticed that the memory usage isn't really growing. Either it need a specific usage pattern to trigger a sudden memory leak, or the bug has been fixed in `1.5.0-beta.14`. I'll keep using the app normally under the profiler and see how it goes

**Comment by sifmelcara** [comment]
@Elip100 did you delete your comment? I wonder what's the situation on your side.

On `1.5.0-beta.14` it seems like there is still leak. But now it only leaks ~15MB per hour, instead of ~200MB per hour in `1.5.0-beta.13`.

**Comment by Elip100** [comment]
> [@Elip100](https://github.com/Elip100) did you delete your comment? I wonder what's the situation on your side.
> 
> On `1.5.0-beta.14` it seems like there is still leak. But now it only leaks ~15MB per hour, instead of ~200MB per hour in `1.5.0-beta.13`.

It's definitely slower but it's still leaking. It's slow enough that I reboot often enough for it to not affect me that much.

**Comment by Kathund** [comment]
> [@Elip100](https://github.com/Elip100) did you delete your comment? I wonder what's the situation on your side.
> 
> On `1.5.0-beta.14` it seems like there is still leak. But now it only leaks ~15MB per hour, instead of ~200MB per hour in `1.5.0-beta.13`.

I seam to be having the same thing just a big higher scale then 15mb an hour
<img width="1041" height="230" alt="Image" src="https://github.com/user-attachments/assets/3347d975-dc13-47bc-b0b9-0c6e2ce3b4de" />

I'm running `1.5.0-beta.14` from nixpkgs
I think this might also be the highest I have seen

**Comment by Core447** [comment]
eb2bc607a8a541ad7d880140275d479f08ba9692 massively improved this for me

**Comment by Kathund** [comment]
> [eb2bc60](https://github.com/StreamController/StreamController/commit/eb2bc607a8a541ad7d880140275d479f08ba9692) massively improved this for me

Built the 1.5.0-beta.16 from source and it seams to be a lot better

**Comment by Core447** [comment]
> > [eb2bc60](https://github.com/StreamController/StreamController/commit/eb2bc607a8a541ad7d880140275d479f08ba9692) massively improved this for me
> 
> Built the 1.5.0-beta.16 from source and it seams to be a lot better

Great!
